src/core/instances.jl

```diff
@@ -0,0 +1,314 @@
+
+parameters(ci::ComponentInstance) = ci.parameters
+
+
+function getindex(mi::ModelInstance, comp_name::Symbol, datum_name::Symbol)
+    if !(comp_name in keys(mi.components))
+        error("Component does not exist in current model")
+    end
+    
+    comp_inst = compinstance(mi, comp_name)
+    vars = comp_inst.variables
+    pars = comp_inst.parameters
+
+    if datum_name in vars.names
+        which = vars
+    elseif datum_name in pars.names
+        which = pars
+    else
+        error("$datum_name is not a parameter or a variable in component $comp_name.")
+    end
+
+    value = getproperty(which, Val(datum_name))
+    # return isa(value, PklVector) || isa(value, TimestepMatrix) ? value.data : value
+    return isa(value, AbstractTimestepMatrix) ? value.data : value
+end
+
+# """
+#     indexcount(mi::ModelInstance, idx_name::Symbol)
+
+# Returns the size of index `idx_name`` in model instance `mi`.
+# """
+# indexcount(mi::ModelInstance, idx_name::Symbol) = indexcounts(mi.md, idx_name)
+
+"""
+    dim_count(mi::ModelInstance, dim_name::Symbol)
+
+Returns the size of index `dim_name`` in model instance `mi`.
+"""
+dim_count(mi::ModelInstance, dim_name::Symbol) = dim_count(mi.md, dim_name)
+
+dim_key_dict(mi::ModelInstance) = dim_key_dict(mi.md)
+
+dim_value_dict(mi::ModelInstance) = dim_value_dict(mi.md)
+
+# """
+#     indexvalues(mi::ModelInstance, idx_name::Symbol)
+
+# Return the values of index `idx_name` in model instance `mi`.
+# """
+# indexvalues(mi::ModelInstance, idx_name::Symbol) = indexvalues(mi.md, idx_name)
+
+function make_clock(mi::ModelInstance, ntimesteps, time_keys::Vector{Int})
+    start = time_keys[1]
+    stop  = time_keys[min(length(time_keys), ntimesteps)]
+    step  = step_size(time_keys)
+    return Clock(start, step, stop)
+end
+
+function reset_variables(ci::ComponentInstance)
+    # println("reset_variables($(ci.comp_id))")
+    vars = ci.variables
+
+    for (name, ref) in zip(vars.names, vars.types.parameters)
+        # Everything is held in a Ref{}, so get the parameters to that...
+        T = ref.parameters[1]
+        value = getproperty(vars, Val(name))
+
+        if (T <: AbstractTimestepMatrix || T <: AbstractMatrix) && eltype(value) <: AbstractFloat
+            fill!(value, NaN)
+
+        elseif T <: AbstractFloat
+            setproperty!(vars, Val(name), NaN)
+        end
+    end
+end
+
+function init(mi::ModelInstance)
+    for ci in components(mi)
+        init(mi, ci)
+    end
+end
+
+function init(mi::ModelInstance, ci::ComponentInstance)
+    reset_variables(ci)
+
+    comp_def = compdef(mi.md, ci.comp_name)
+
+    if init_expr(comp_def) != nothing
+        module_name = compmodule(ci.comp_id)
+        comp_name = ci.comp_name
+        pars = ci.parameters
+        vars = ci.variables
+        dims = dim_value_dict(mi.md)
+
+        Base.invokelatest(init, (Val(module_name), Val(comp_name), pars, vars, dims)...)
```

Comment:
Uh, are we running into world age problems here? If so, that in my mind is another strong reason to emit the functions in the ``@defcomp`` macro already and avoid all of this complication.

Reply:
Changed to emit the function from the macro and avoid `invokelatest` (in the upcoming commit.)